tensorflow/c/kernels_experimental.cc

```diff
@@ -285,6 +288,93 @@ void TF_AssignUpdateVariable(TF_OpKernelContext* ctx, int input_index,
   TF_SetStatus(tf_status, TF_OK, "");
 }
 
+struct TmpVar : public ResourceBase {
+  tensorflow::mutex mu;
+  Tensor val;
+  std::string name;
+  std::string DebugString() const { return name; }
+  ~TmpVar() override { VLOG(3) << "TmpVar " << name << " deleted"; }
+};
+
+// Makes a unique name for a temporary variable inside a while loop body,
+//  because loop can be executed in multiple iterations in parallel.
+std::string TemporaryVariableName(
+    const std::string& var_name,
+    const tensorflow::FrameAndIter& control_frame) {
+  if (control_frame.frame_id != tensorflow::kIllegalFrameId &&
+      control_frame.iter_id != tensorflow::kIllegalIterId) {
+    return tensorflow::strings::StrCat(var_name,
+                                       "/frame:", control_frame.frame_id,
+                                       "/iter:", control_frame.iter_id);
+  }
+  return var_name;
+}
+
+void TF_TemporaryVariable(TF_OpKernelContext* ctx, TF_DataType dtype,
+                          const int64_t* dims, int num_dims,
+                          TF_StringView* var_name,
+                          void (*allocFunc)(TF_OpKernelContext*, TF_Tensor*,
+                                            TF_DataType, const int64_t*, int,
+                                            TF_Status*),
+                          TF_Status* tf_status) {
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  tensorflow::ResourceMgr* rm = context->resource_manager();
+  OP_REQUIRES(context, rm,
+              tensorflow::errors::Internal("No per-step resource manager."));
+
+  std::string unique_name =
+      TemporaryVariableName(var_name->data, context->frame_iter());
+  auto* tmp_var = new TmpVar;
+  OP_REQUIRES(
+      context, tmp_var,
+      tensorflow::errors::ResourceExhausted("Could not allocate TmpVar."));
+  tmp_var->name = unique_name;
+
+  Status s;
+  std::unique_ptr<TF_Tensor, decltype(&TF_DeleteTensor)> tmp_var_tf(
+      tensorflow::TF_TensorFromTensor(tmp_var->val, &s), TF_DeleteTensor);
+  OP_REQUIRES_OK(context, s);
+  allocFunc(ctx, tmp_var_tf.get(), dtype, dims, num_dims, tf_status);
+  s = tensorflow::StatusFromTF_Status(tf_status);
+  if (!s.ok()) tmp_var->Unref();
+  OP_REQUIRES_OK(context, s);
+
+  OP_REQUIRES_OK(context, TF_TensorToTensor(tmp_var_tf.get(), &tmp_var->val));
+  OP_REQUIRES_OK(context,
+                 context->step_container()->Create(rm, unique_name, tmp_var));
+  context->set_output_ref(0, &tmp_var->mu, &tmp_var->val);
+
+  if (context->track_allocations()) {
+    context->record_persistent_memory_allocation(tmp_var->val.AllocatedBytes());
+  }
+
+  TF_SetStatus(tf_status, TF_OK, "");
+}
+
+void TF_DestroyTemporaryVariable(TF_OpKernelContext* ctx, const int index,
+                                 TF_StringView* var_name,
+                                 TF_Status* tf_status) {
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  CHECK(IsRefType(context->input_dtype(0)));
```

Comment:
Nit: We don't use `CHECK` or `DCHECK` anymore (in new code).

Reply:
@penpornk do you have any suggestion to replace CHECK? I saw class DestroyTemporaryVariableOp still use CHECK. thanks
```
class DestroyTemporaryVariableOp : public OpKernel {
144  public:
145   explicit DestroyTemporaryVariableOp(OpKernelConstruction* context)
146       : OpKernel(context) {
147     OP_REQUIRES(context, IsRefType(context->input_type(0)),
148                 errors::InvalidArgument("lhs input needs to be a ref type"));
149     OP_REQUIRES_OK(context, context->GetAttr("var_name", &var_name_));
150     OP_REQUIRES(context, !var_name_.empty(),
151                 errors::InvalidArgument("Missing var_name attribute"));
152   }
153
154   void Compute(OpKernelContext* context) override {
155     // NOTE(pbar): All other mutators of the Tensor Ref *must* have completed
156     // their execution before this DestroyTemporaryVariable op executes.
157     // This is typically achieved using control dependencies.
158     CHECK(IsRefType(context->input_dtype(0)));

```

Reply:
Re:

>  I saw class DestroyTemporaryVariableOp still use CHECK. 

Forgot to clarify why you saw other code still using CHECKs. That's because CHECK/DCHECKs were used in so many places that it would take a lot of work to replace them all. So we only enforce this requirements to new code, i.e., we try not to add anymore CHECK/DCHECKs in new code. :)

Reply:
@penpornk  I saw it in variable_ops.cc, thanks.
https://github.com/tensorflow/tensorflow/blob/b08d6e740c2cb002409d695ccae4f9517402107c/tensorflow/core/kernels/variable_ops.cc#L158

Reply:
@penpornk  done. Thanks